pallets/dapp-staking-v3/src/types.rs

```diff
@@ -638,4 +810,463 @@ impl EraInfo {
+///
+#[derive(Encode, Decode, MaxEncodedLen, Copy, Clone, Debug, PartialEq, Eq, TypeInfo)]
+pub struct ContractStakingInfo {
+    #[codec(compact)]
+    vp_staked_amount: Balance,
+    #[codec(compact)]
+    bep_staked_amount: Balance,
+    #[codec(compact)]
+    era: EraNumber,
+    #[codec(compact)]
+    period: PeriodNumber,
+}
+
+impl ContractStakingInfo {
+    /// Create new instance of `ContractStakingInfo` with specified era & period.
+    /// These parameters are immutable.
+    ///
+    /// Staked amounts are initialized to zero and can be increased or decreased.
+    pub fn new(era: EraNumber, period: PeriodNumber) -> Self {
+        Self {
+            vp_staked_amount: Balance::zero(),
+            bep_staked_amount: Balance::zero(),
+            era,
+            period,
+        }
+    }
+
+    /// Total staked amount on the contract.
+    pub fn total_staked_amount(&self) -> Balance {
+        self.vp_staked_amount.saturating_add(self.bep_staked_amount)
+    }
+
+    /// Staked amount of the specified period type.
+    ///
+    /// Note:
+    /// It is possible that voting period stake is reduced during the build&earn period.
+    /// This is because stakers can unstake their funds during the build&earn period, which can
+    /// chip away from the voting period stake.
+    pub fn staked_amount(&self, period_type: PeriodType) -> Balance {
+        match period_type {
+            PeriodType::Voting => self.vp_staked_amount,
+            PeriodType::BuildAndEarn => self.bep_staked_amount,
+        }
+    }
+
+    /// Era for which this entry is relevant.
+    pub fn era(&self) -> EraNumber {
+        self.era
+    }
+
+    /// Period for which this entry is relevant.
+    pub fn period(&self) -> PeriodNumber {
+        self.period
+    }
+
+    /// Stake specified `amount` on the contract, for the specified `period_type`.
+    pub fn stake(&mut self, amount: Balance, period_type: PeriodType) {
+        match period_type {
+            PeriodType::Voting => self.vp_staked_amount.saturating_accrue(amount),
+            PeriodType::BuildAndEarn => self.bep_staked_amount.saturating_accrue(amount),
+        }
+    }
+
+    /// Unstake specified `amount` from the contract, for the specified `period_type`.
+    pub fn unstake(&mut self, amount: Balance, period_type: PeriodType) {
+        match period_type {
+            PeriodType::Voting => self.vp_staked_amount.saturating_reduce(amount),
+            PeriodType::BuildAndEarn => {
+                let overflow = amount.saturating_sub(self.bep_staked_amount);
+                self.bep_staked_amount.saturating_reduce(amount);
+                self.vp_staked_amount.saturating_reduce(overflow);
+            }
+        }
+    }
+
+    /// `true` if no stake exists, `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.vp_staked_amount.is_zero() && self.bep_staked_amount.is_zero()
+    }
+}
+
+const STAKING_SERIES_HISTORY: u32 = 3;
```

Comment:
From the technical doc, I'm assuming this'll be configurable therefore used as generic parameter below.
If not, the should be hardcoded.

Reply:
I see, I got confused over many other `history depth`.

If we need to cover this for only three elements (and three scenarios), then why use `BoundedVec`?
I mean, can't we simply use three struct variables like `s0, s1, s2` ?

Reply:
Let me answer with a thought (or code?) experiment - how would you implement `stake` or `unstake` using 3 separate attributes? How would that code look like?

If I'm not mistaken, it would be if within if within if, etc.